[image-picker][android] refactor: Add custom cropping activity with theme support

hirbod opened the pull request:
# Why

The image picker's cropping interface currently doesn't adapt to the app's theme, meaning it doesn't properly support dark mode or light mode theming. This creates an inconsistent user experience where the cropping UI doesn't match the rest of the app's appearance.

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

# How

I implemented a custom cropping activity that wraps the existing CropImageActivity to provide proper theme support. The solution uses reflection to access the underlying library's private fields and methods, allowing us to apply custom theming without forking the library. I added customizable color attributes that developers can use to integrate the cropping interface with their app's theme, and the activity automatically detects whether the app is in light or dark mode to apply appropriate colors for the toolbar, icons, and status bar. The implementation maintains full backward compatibility while providing a seamless, theme-aware cropping experience.

We could consider extending the config plugin to support styles.xml and values-night/styles.xml, although it’s already looking pretty decent imo. Alternatively, we could just update the docs (cc @amandeepmittal) to include some tips about this.

```xml
    <attr name="expoCropToolbarColor" format="color"/>
    <attr name="expoCropToolbarIconColor" format="color"/>
    <attr name="expoCropToolbarActionTextColor" format="color"/>
    <attr name="expoCropBackButtonIconColor" format="color"/>
    <attr name="expoCropBackgroundColor" format="color"/>
```

<!--
How did you build this feature or fix this bug and why?
-->

# Test Plan

Bare Expo

| Before | After |
|--------|-------|
| <img width="485" alt="Before" src="https://github.com/user-attachments/assets/82650a8a-ec47-4f7d-836a-8e232ec1d09d" /> | <img width="485" alt="After" src="https://github.com/user-attachments/assets/77ef53d5-284c-468e-81c5-9bfedbb87904" /> |


https://github.com/user-attachments/assets/ce27b6f9-3adc-46e8-8121-e0d8f3055697



<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

hirbod committed:
[image-picker] Refactor cropping activity and update manifest

- Replaced `CropImageActivity` with `ExpoCropImageActivity` in the AndroidManifest.xml for custom theming.
- Introduced `ExpoCropImageActivity` to apply a custom color palette based on the application's theme.
- Updated `CropImageContract` to use the new `ExpoCropImageActivity`.
- Added new attributes in `attrs.xml` for customizable colors in the cropping UI.

hirbod committed:
update changelog

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ExpoCropImageActivity.kt:
```diff
@@ -0,0 +1,110 @@
+      if (theme.resolveAttribute(attr, tv, true)) tv.data else null
+    }.getOrNull()
+
+    val customToolbar = getThemeColor(R.attr.expoCropToolbarColor)
+    val customIconColor = getThemeColor(R.attr.expoCropToolbarIconColor)
+    val customActionTextColor = getThemeColor(R.attr.expoCropToolbarActionTextColor)
+    val customBackButtonIconColor = getThemeColor(R.attr.expoCropBackButtonIconColor)
+    val customBg = getThemeColor(R.attr.expoCropBackgroundColor)
+
+    if (isNight) {
+      opts.activityBackgroundColor = customBg ?: Color.BLACK
+      opts.toolbarColor = customToolbar ?: Color.BLACK
+      val toolbarWidgetColor = customIconColor ?: Color.WHITE
+      opts.toolbarTitleColor = toolbarWidgetColor
+      opts.toolbarBackButtonColor = customBackButtonIconColor ?: toolbarWidgetColor
+      opts.activityMenuIconColor = toolbarWidgetColor
+      opts.activityMenuTextColor = customActionTextColor ?: Color.WHITE
+      currentIconColor = toolbarWidgetColor
+      window.statusBarColor = opts.toolbarColor ?: Color.BLACK
+      WindowInsetsControllerCompat(window, window.decorView).isAppearanceLightStatusBars = false
+    } else {
+      opts.activityBackgroundColor = customBg ?: Color.WHITE
+      opts.toolbarColor = customToolbar ?: opts.activityBackgroundColor ?: Color.WHITE
+      val toolbarWidgetColor = customIconColor ?: Color.BLACK
+      opts.toolbarTitleColor = toolbarWidgetColor
+      opts.toolbarBackButtonColor = customBackButtonIconColor ?: toolbarWidgetColor
+      opts.activityMenuIconColor = toolbarWidgetColor
+      opts.activityMenuTextColor = customActionTextColor ?: Color.BLACK
+      currentIconColor = toolbarWidgetColor
+      window.statusBarColor = opts.toolbarColor ?: Color.WHITE
+      WindowInsetsControllerCompat(window, window.decorView).isAppearanceLightStatusBars = true
+    }
```
```suggestion
		fun CropImageOptions.applyPallet(defualtColor: Color) {
      val customToolbar = getThemeColor(R.attr.expoCropToolbarColor)
      val customIconColor = getThemeColor(R.attr.expoCropToolbarIconColor)
      val customActionTextColor = getThemeColor(R.attr.expoCropToolbarActionTextColor)
      val customBackButtonIconColor = getThemeColor(R.attr.expoCropBackButtonIconColor)
      val customBg = getThemeColor(R.attr.expoCropBackgroundColor)

	    val toolbarWidgetColor = customIconColor ?: defaultColor
  			
		  activityBackgroundColor = customBg ?: defaultColor
	    toolbarColor = customToolbar ?: defaultColor
	    toolbarTitleColor = toolbarWidgetColor
	    toolbarBackButtonColor = customBackButtonIconColor ?: toolbarWidgetColor
	    activityMenuIconColor = toolbarWidgetColor
	    activityMenuTextColor = customActionTextColor ?: Color.WHITE
  
	    /* Those lines need to be extracted from this function 
	      currentIconColor = toolbarWidgetColor 
	      window.statusBarColor = opts.toolbarColor ?: Color.BLACK
        WindowInsetsControllerCompat(window, window.decorView).isAppearanceLightStatusBars = false
     */
     }
```
Could you extract that logic to a separate function? My suggestion is just an example of how the implementation could be done. It's not completed!

Also, I wanted to confirm that `toolbarColor` color is set correctly, because in the dark mode it's set to `customToolbar ?: Color.BLACK` but in the light mode to `customToolbar ?: opts.activityBackgroundColor ?: Color.WHITE`. Why did we end up skipping `activityBackgroundColor` in the dark mode?

hirbod replied:
Thanks for flagging, definitely an oversight. I wrested for a very long time to understand the theming and to make the the colors reactive. Looks like a leftover, will adjust.

hirbod committed:
[image-picker] Enhance cropping UI theming and utility functions

- Introduced `ExpoCropImageUtils` for managing color theming in `ExpoCropImageActivity`.
- Refactored `applyPalette` method to streamline color application based on theme attributes.
- Added new color resources for light and dark modes in `colors.xml` and `colors-night.xml`.
- Removed unused imports in `CropImageContract` to clean up the codebase.

hirbod commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ExpoCropImageActivity.kt:
```diff
@@ -0,0 +1,110 @@
+    fun getThemeColor(attr: Int): Int? = runCatching {
+      val tv = TypedValue()
+      if (theme.resolveAttribute(attr, tv, true)) tv.data else null
+    }.getOrNull()
```
Moved to Utils

amandeepmittal commented:
> We could consider extending the config plugin to support styles.xml and values-night/styles.xml, although it’s already looking pretty decent imo. Alternatively, we could just update the docs (cc @amandeepmittal) to include some tips about this.

I am not sure what's the best solution here but if you consider adding info in docs, let's add a minimal example that a developer can run on their end within the Image Picker reference section.

hirbod commented:
@amandeepmittal We can address that later if there’s actual demand. I think white and black are solid defaults for a cropper. If needed, we can explain how to customize them or provide a config plugin.
